To add <>/target/classes directory to tests instrumentation classpath?

I noticed ambiguous behavior of atom-maven-plugin, instrumenting the folloing test (from project [oo-data](https://github.com/pragmaticobjects/oo-data)):

```
@NotAtom
public class SrcFileJavaPoetTest extends TestsSuite {
    private static final DeclExplicit<?> COM_PRAGMATICOBJECTS_TEST = new DeclExplicit(
        () -> Annotation.class,
        "com.pragmaticobjects.test"
    );
    
    /**
     * Ctor.
     */
    public SrcFileJavaPoetTest() {
        super(
            new TestCase(
                "simplest java source file generation passes",
                new AssertAssumingTemporaryDirectory(
                    tmpDir -> new AssertSourceFileGenerated(
                        new SrcFileJavaPoet(
                            COM_PRAGMATICOBJECTS_TEST,
                            () -> TypeSpec.interfaceBuilder("Test").build(),
                            new DestToPath(tmpDir)
                        ), 
                        tmpDir.resolve("com/pragmaticobjects/test/Test.java"),
                        String.join(
                            System.lineSeparator(),
                            "package com.pragmaticobjects.test;",
                            "",
                            "interface Test {",
                            "}",
                            ""
                        )
                    )
                )
            )
        );
    }
}
```

Instrumenter status:
```
java.home: /home/skapral/Software/jdk-9.0.4_linux-x64_bin/jdk-9.0.4
java.class.path:
 /home/skapral/.m2/repository/io/vavr/vavr/0.9.2/vavr-0.9.2.jar
 /home/skapral/.m2/repository/io/vavr/vavr-match/0.9.2/vavr-match-0.9.2.jar
 /home/skapral/.m2/repository/com/squareup/javapoet/1.11.1/javapoet-1.11.1.jar
 /home/skapral/.m2/repository/org/apache/commons/commons-text/1.4/commons-text-1.4.jar
 /home/skapral/.m2/repository/org/apache/commons/commons-lang3/3.7/commons-lang3-3.7.jar
 /home/skapral/.m2/repository/commons-io/commons-io/1.3.2/commons-io-1.3.2.jar
 /home/skapral/.m2/repository/com/pragmaticobjects/oo/tests/tests-junit5/0.0.0/tests-junit5-0.0.0.jar
 /home/skapral/.m2/repository/com/pragmaticobjects/oo/tests/tests-core/0.0.0/tests-core-0.0.0.jar
 /home/skapral/.m2/repository/org/junit/jupiter/junit-jupiter-engine/5.2.0/junit-jupiter-engine-5.2.0.jar
 /home/skapral/.m2/repository/org/apiguardian/apiguardian-api/1.0.0/apiguardian-api-1.0.0.jar
 /home/skapral/.m2/repository/org/junit/platform/junit-platform-engine/1.2.0/junit-platform-engine-1.2.0.jar
 /home/skapral/.m2/repository/org/junit/platform/junit-platform-commons/1.2.0/junit-platform-commons-1.2.0.jar
 /home/skapral/.m2/repository/org/opentest4j/opentest4j/1.1.0/opentest4j-1.1.0.jar
 /home/skapral/.m2/repository/org/junit/jupiter/junit-jupiter-api/5.2.0/junit-jupiter-api-5.2.0.jar
 /home/skapral/.m2/repository/org/assertj/assertj-core/3.11.1/assertj-core-3.11.1.jar
 /home/skapral/.m2/repository/org/mockito/mockito-core/2.23.4/mockito-core-2.23.4.jar
 /home/skapral/.m2/repository/net/bytebuddy/byte-buddy/1.9.3/byte-buddy-1.9.3.jar
 /home/skapral/.m2/repository/net/bytebuddy/byte-buddy-agent/1.9.3/byte-buddy-agent-1.9.3.jar
 /home/skapral/.m2/repository/org/objenesis/objenesis/2.6/objenesis-2.6.jar
 /home/skapral/GitHub/oo-data/data-core/target/test-classes
```

Exception:
```
Failed to execute goal com.pragmaticobjects.oo.atom:atom-maven-plugin:0.0.15:instrument-tests (default) on project data-core: Execution default of goal com.pragmaticobjects.oo.atom:atom-maven-plugin:0.0.15:instrument-tests failed: java.lang.IllegalStateException: Cannot resolve type description for com.pragmaticobjects.oo.data.model.declaration.DeclExplicit -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.pragmaticobjects.oo.atom:atom-maven-plugin:0.0.15:instrument-tests (default) on project data-core: Execution default of goal com.pragmaticobjects.oo.atom:atom-maven-plugin:0.0.15:instrument-tests failed: java.lang.IllegalStateException: Cannot resolve type description for com.pragmaticobjects.oo.data.model.declaration.DeclExplicit
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default of goal com.pragmaticobjects.oo.atom:atom-maven-plugin:0.0.15:instrument-tests failed: java.lang.IllegalStateException: Cannot resolve type description for com.pragmaticobjects.oo.data.model.declaration.DeclExplicit
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:145)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.lang.RuntimeException: java.lang.IllegalStateException: Cannot resolve type description for com.pragmaticobjects.oo.data.model.declaration.DeclExplicit
	at com.pragmaticobjects.oo.atom.codegen.stage.ByteBuddyStage.apply(ByteBuddyStage.java:82)
	at com.pragmaticobjects.oo.atom.codegen.stage.SequenceStage.apply(SequenceStage.java:61)
	at com.pragmaticobjects.oo.atom.instrumentation.ApplyStages.apply(ApplyStages.java:88)
	at com.pragmaticobjects.oo.atom.maven.BaseMojo.doInstrumentation(BaseMojo.java:51)
	at com.pragmaticobjects.oo.atom.maven.InstrumentTestsMojo.execute(InstrumentTestsMojo.java:51)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	... 21 more
Caused by: java.lang.IllegalStateException: Cannot resolve type description for com.pragmaticobjects.oo.data.model.declaration.DeclExplicit
	at net.bytebuddy.pool.TypePool$Resolution$Illegal.resolve(TypePool.java:134)
	at net.bytebuddy.pool.TypePool$Default$LazyTypeDescription$GenericTypeToken$ForParameterizedType$LazyParameterizedType.getOwnerType(TypePool.java:4911)
	at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor.onParameterizedType(TypeDescription.java:1576)
	at net.bytebuddy.description.type.TypeDescription$Generic$Visitor$Substitutor$ForDetachment.onParameterizedType(TypeDescription.java:1731)
	at net.bytebuddy.description.type.TypeDescription$Generic$OfParameterizedType.accept(TypeDescription.java:4295)
	at net.bytebuddy.description.type.TypeDescription$Generic$LazyProjection.accept(TypeDescription.java:5210)
	at net.bytebuddy.description.field.FieldDescription$AbstractBase.asToken(FieldDescription.java:143)
	at net.bytebuddy.description.field.FieldDescription$AbstractBase.asToken(FieldDescription.java:87)
	at net.bytebuddy.description.field.FieldList$AbstractBase.asTokenList(FieldList.java:47)
	at net.bytebuddy.dynamic.scaffold.InstrumentedType$Factory$Default$1.represent(InstrumentedType.java:222)
	at net.bytebuddy.ByteBuddy.redefine(ByteBuddy.java:698)
	at com.pragmaticobjects.oo.atom.codegen.stage.ByteBuddyStage.apply(ByteBuddyStage.java:74)
	... 26 more
```

Probably root cause is that "/target/classes" directory was not added to class path. 

## Workaround
The atomizer passes successfully if static constant is eliminated:

```
public class SrcFileJavaPoetTest extends TestsSuite {
    /**
     * Ctor.
     */
    public SrcFileJavaPoetTest() {
        super(
            new TestCase(
                "simplest java source file generation passes",
                new AssertAssumingTemporaryDirectory(
                    tmpDir -> new AssertSourceFileGenerated(
                        new SrcFileJavaPoet(
                            new DeclExplicit(
                                () -> Annotation.class,
                                "com.pragmaticobjects.test"
                            ),
                            () -> TypeSpec.interfaceBuilder("Test").build(),
                            new DestToPath(tmpDir)
                        ), 
                        tmpDir.resolve("com/pragmaticobjects/test/Test.java"),
                        String.join(
                            System.lineSeparator(),
                            "package com.pragmaticobjects.test;",
                            "",
                            "interface Test {",
                            "}",
                            ""
                        )
                    )
                )
            )
        );
    }
}
```

--- a/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/BaseMojo.java
+++ b/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/BaseMojo.java
@@ -6,6 +6,7 @@ import com.pragmaticobjects.oo.atom.codegen.cp.ClassPath;
 import com.pragmaticobjects.oo.atom.codegen.cp.CpFromString;
 import com.pragmaticobjects.oo.atom.codegen.stage.Stage;
 import com.pragmaticobjects.oo.atom.instrumentation.ApplyStages;
+import io.vavr.collection.HashSet;
 import org.apache.maven.artifact.Artifact;
 import org.apache.maven.plugin.AbstractMojo;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -17,6 +18,7 @@ import java.io.File;
 import java.nio.file.Path;
 import java.util.Optional;
 import java.util.stream.Collectors;
+import org.apache.maven.plugin.descriptor.PluginDescriptor;
 
 /**
  * Base Mojo for different instrumentations
@@ -31,13 +33,19 @@ public abstract class BaseMojo extends AbstractMojo {
     @Parameter
     private String[] excludePackages;
 
-    protected final void doInstrumentation(Stage stage, Path workingDirectory) throws MojoExecutionException, MojoFailureException {
-        if(!project.getPackaging().equals("pom") || instrumentPomProjects) {
-            String classPath = project.getArtifacts().stream()
+    protected final ClassPath buildClassPath() {
+        PluginDescriptor pluginDescriptor = (PluginDescriptor) this.getPluginContext().get("pluginDescriptor");
+        String classPath = HashSet.ofAll(project.getArtifacts())
+                    .addAll(pluginDescriptor.getArtifacts())
                     .map(Artifact::getFile)
                     .map(File::toString)
                     .collect(Collectors.joining(":"));
-            ClassPath cp = new CpFromString(classPath);
+        ClassPath cp = new CpFromString(classPath);
+        return cp;
+    }
+    
+    protected final void doInstrumentation(Stage stage, ClassPath cp, Path workingDirectory) throws MojoExecutionException, MojoFailureException {
+        if(!project.getPackaging().equals("pom") || instrumentPomProjects) {
             new ApplyStages(
                 cp,
                 workingDirectory,

--- a/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/InstrumentMojo.java
+++ b/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/InstrumentMojo.java
@@ -50,6 +50,7 @@ public class InstrumentMojo extends BaseMojo {
     public final void execute() throws MojoExecutionException, MojoFailureException {
         doInstrumentation(
             new StandardInstrumentationStage(stubbedInstrumentation),
+            buildClassPath(),
             Paths.get(outputDirectory)
         );
     }

--- a/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/InstrumentTestsMojo.java
+++ b/atom-maven-plugin/src/main/java/com/pragmaticobjects/oo/atom/maven/InstrumentTestsMojo.java
@@ -23,6 +23,8 @@
  */
 package com.pragmaticobjects.oo.atom.maven;
 
+import com.pragmaticobjects.oo.atom.codegen.cp.CpCombined;
+import com.pragmaticobjects.oo.atom.codegen.cp.CpFromString;
 import com.pragmaticobjects.oo.atom.codegen.stage.StandardInstrumentationStage;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.MojoFailureException;
@@ -40,8 +42,10 @@ import java.nio.file.Paths;
  */
 @Mojo(name = "instrument-tests", defaultPhase = LifecyclePhase.PROCESS_TEST_CLASSES, requiresDependencyResolution = ResolutionScope.TEST)
 public class InstrumentTestsMojo extends BaseMojo {
-    @Parameter(defaultValue = "${project.build.testOutputDirectory}", required = true, readonly = true)
+    @Parameter(defaultValue = "${project.build.outputDirectory}", required = true, readonly = true)
     protected String outputDirectory;
+    @Parameter(defaultValue = "${project.build.testOutputDirectory}", required = true, readonly = true)
+    protected String testOutputDirectory;
 
     @Parameter(defaultValue = "false", required = true, readonly = true)
     protected boolean stubbedInstrumentation;
@@ -50,7 +54,13 @@ public class InstrumentTestsMojo extends BaseMojo {
     public void execute() throws MojoExecutionException, MojoFailureException {
         doInstrumentation(
             new StandardInstrumentationStage(stubbedInstrumentation),
-            Paths.get(outputDirectory)
+            new CpCombined(
+                buildClassPath(),
+                new CpFromString(
+                    outputDirectory
+                )
+            ),
+            Paths.get(testOutputDirectory)
         );
     }
 }

